### PR TITLE
ci(feat): refactor error handling

### PR DESCRIPTION
**Pull request by joshmarinacci**
**Description**:

Create a common set-failure-mode workflow that can be used to set the failure mode variable of a single task, or aggregate a bunch of child tasks.

**Related issue(s)**:

Partially addresses https://github.com/hiero-ledger/hiero-consensus-node/issues/25469?issue=hiero-ledger%7Chiero-consensus-node%7C25509 and https://github.com/hiero-ledger/hiero-consensus-node/issues/25513 and https://github.com/hiero-ledger/hiero-consensus-node/issues/25469?issue=hiero-ledger%7Chiero-consensus-node%7C25470

**Notes for reviewer**:

**Checklist**

- [ ] Documented (Code comments, README, etc.)
- [ ] Tested (unit, integration, etc.)

### CODE REVIEW

File: .github/workflows/zxc-verify-docker-build-determinism.yaml [hunk 1]
```diff
@@ -438,36 +438,24 @@ jobs:
     if: ${{ !cancelled() && always() }}
     outputs:
       failure-mode: ${{ steps.set-failure-mode.outputs.failure-mode }}
+      failed-tests: ${{ steps.set-failure-mode.outputs.failed-tests }}
     steps:
-      - name: Harden Runner
-        uses: step-security/harden-runner@8d3c67de8e2fe68ef647c8db1e6a09f647780f40 # v2.19.0
+      - name: Initialize GitHub Job
+        uses: PandasWhoCode/initialize-github-job@d93cd457947758129c6d93e09f89ad528119960f # v1.0.8
```
[the remark this answers is not in the document]
Comment (andrewb1269) [reply .github/workflows/zxc-verify-docker-build-determinism.yaml]: This is a valid finding.

File: .github/workflows/support/scripts/set-failure-mode.sh [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+  if [[ "$status" == "failure" ]]; then
+    [[ "$mode" == "cancelled" ]] && mode="workflow"
+    case "$mode" in
+      workflow) failure_mode="workflow" ;;
+      test) [[ "$failure_mode" != "workflow" ]] && failure_mode="test" ;;
+    esac
+    failed_tests="${failed_tests:+${failed_tests}, }${name}"
+  elif [[ "$status" == "cancelled" ]]; then
+    [[ "$failure_mode" == "none" ]] && failure_mode="cancelled"
+    failed_tests="${failed_tests:+${failed_tests}, }${name} (cancelled)"
```
[the remark this answers is not in the document]
Comment (andrewb1269) [reply .github/workflows/support/scripts/set-failure-mode.sh]: This is a valid finding.

File: .github/workflows/zxc-verify-docker-build-determinism.yaml (hunk 1 above)
[the remark this answers is not in the document]
Comment (joshmarinacci) [reply .github/workflows/zxc-verify-docker-build-determinism.yaml]: fixed.

File: .github/workflows/support/scripts/set-failure-mode.sh [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+
+read -ra test_arr <<< "${INPUT_TEST_OUTCOMES}"
+read -ra workflow_arr <<< "${INPUT_WORKFLOW_OUTCOMES}"
+
+if has_failure "${test_arr[@]}"; then
+  failure_mode="test"
+fi
+if has_failure "${workflow_arr[@]}"; then
+  failure_mode="workflow"
+fi
```
Comment (andrewb1269) [inline .github/workflows/support/scripts/set-failure-mode.sh]: If there is a test failure and a workflow failure, then the test failure is overwritten by the workflow failure. Suggest changing to this to `elif` so if we have a test failure we do not worry about the workflow failure(s).

File: .github/workflows/support/scripts/set-failure-mode.sh [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+
+read -ra job_status_arr <<< "${INPUT_JOB_STATUSES}"
+read -ra job_failure_mode_arr <<< "${INPUT_JOB_FAILURE_MODES}"
+read -ra job_names_arr <<< "${INPUT_JOB_NAMES}"
+
+for i in "${!job_status_arr[@]}"; do
+  status="${job_status_arr[$i]}"
+  mode="${job_failure_mode_arr[$i]:-none}"
+  name="${job_names_arr[$i]:-job-$i}"
```
Comment (andrewb1269) [inline .github/workflows/support/scripts/set-failure-mode.sh]: Can we add some comments to this processing? I'm unfamiliar with these array operations.

File: .github/workflows/support/scripts/set-failure-mode.sh [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+
+  if [[ "$status" == "failure" ]]; then
+    [[ "$mode" == "cancelled" ]] && mode="workflow"
+    case "$mode" in
+      workflow) failure_mode="workflow" ;;
+      test) [[ "$failure_mode" != "workflow" ]] && failure_mode="test" ;;
```
Comment (andrewb1269) [inline .github/workflows/support/scripts/set-failure-mode.sh]: This will overwrite the `failure_mode` to `workflow` and hide potential `test` failure types. We want to prioritize `test` failures over `workflow` failures.

File: .github/workflows/support/scripts/set-failure-mode.sh [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+    esac
+    failed_tests="${failed_tests:+${failed_tests}, }${name}"
```
Comment (andrewb1269) [inline .github/workflows/support/scripts/set-failure-mode.sh:60]: Can we add a comment and make this line more readable?

File: .github/workflows/support/scripts/set-failure-mode.sh [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+  elif [[ "$status" == "cancelled" ]]; then
+    [[ "$failure_mode" == "none" ]] && failure_mode="cancelled"
```
Comment (andrewb1269) [inline .github/workflows/support/scripts/set-failure-mode.sh]: `failure_mode` can be `workflow`, `test`, or `none`. It should stay as `none` since there was not a failure, it was only a cancelled workflow.

File: .github/workflows/support/scripts/set-failure-mode.sh (hunk 2 above)
Comment (andrewb1269) [inline .github/workflows/support/scripts/set-failure-mode.sh]: We don't know if it was a failed test or not, so we should not add it to the `failed_tests` list.

File: .github/workflows/support/scripts/set-failure-mode.sh [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+
+if has_failure "${test_arr[@]}"; then
+  failure_mode="test"
+fi
+if has_failure "${workflow_arr[@]}"; then
+  failure_mode="workflow"
+fi
+if [[ "${failure_mode}" == "none" ]]; then
+  if has_cancellation "${workflow_arr[@]}" || has_cancellation "${test_arr[@]}"; then
+    failure_mode="cancelled"
```
Comment (andrewb1269) [inline .github/workflows/support/scripts/set-failure-mode.sh]: Why are we adding a `failure_mode` type of `cancelled`?

**Comment by andrewb1269** [comment]
We are trying to provide a reason for why XTS workflows are cancelled to our Slack channels. I suggest we provide it as a separate variable, rather than trying to include it into the `failure_mode` variable. If we provide a separate variable, we can populate that variable only when there is a case to check.

File: .github/workflows/zxc-xts-tests.yaml [hunk 9]
```diff
@@ -385,93 +385,41 @@ jobs:
...
+            xts-timing-sensitive-tests
+            xts-hammer-tests
+            xts-hapi-tests
+            xts-otter-tests
+            hedera-node-jrs-panel
+            sdk-tck-regression-panel
+            mirror-node-regression-panel
+            json-rpc-relay-regression-panel
+            block-node-regression-panel
+        run: bash .github/workflows/support/scripts/set-failure-mode.sh
```
[the remark this answers is not in the document]
Comment (andrewb1269) [reply .github/workflows/zxc-xts-tests.yaml:493]: The initialize github job does provide a checkout of the repository.

**Review by andrewb1269 — CHANGES_REQUESTED** [review]

**Comment by andrewb1269** [comment]
The remaining comments are regarding:
1. Cancelled tests being processed vs. not being processed. I'd propose we do NOT use cancelled tests in this PR and can discuss for a follow-on PR.
2. The priority of `test` failures vs. `workflow` failures. I think the order should be: `test` > `workflow` > `none`, then we will add `cancelled` in a future PR giving us `test` > `workflow` > `cancelled` > `none`

**Review by andrewb1269 — DISMISSED** [review]

File: .github/workflows/zxc-xts-tests.yaml [hunk 10]
```diff
@@ -403,98 +403,49 @@ jobs:
...
-            "hedera-node-jrs-panel"
-            "sdk-tck-regression-panel"
-            "mirror-node-regression-panel"
-            "json-rpc-relay-regression-panel"
-            "block-node-regression-panel"
+            test
+            test
+            test
+            test
+            test
```
Comment (andrewb1269) [inline .github/workflows/zxc-xts-tests.yaml]: Is it expected to have all these `test` items here? I think they need to be changed back to the panel names instead.
```suggestion
            "hedera-node-jrs-panel"
            "sdk-tck-regression-panel"
            "mirror-node-regression-panel"
            "json-rpc-relay-regression-panel"
            "block-node-regression-panel"
```

**Review by andrewb1269 — CHANGES_REQUESTED** [review]

File: .github/workflows/zxc-xts-tests.yaml [hunk 11]
```diff
@@ -403,98 +403,50 @@ jobs:
...
-          done
-
-          echo "failure-mode=${failure_mode}" >> "${GITHUB_OUTPUT}"
-          echo "failed-tests=${failed_tests}" >> "${GITHUB_OUTPUT}"
+            test
+            test
+            test
+            test
+            test
+            test
```
Comment (andrewb1269) [inline .github/workflows/zxc-xts-tests.yaml]: Looks like there's still this block of `test` items. I think this should be changed.

**Review by andrewb1269 — CHANGES_REQUESTED** [review]

File: .github/workflows/zxc-xts-tests.yaml (hunk 11 above)
Comment (andrewb1269) [reply .github/workflows/zxc-xts-tests.yaml]: The workflows that do not emit a failure mode should be updated to align with our required workflows.

When complete, this section should be:
```suggestion
            ${{ needs.hedera-node-jrs-panel.outputs.failure-mode || 'none' }}
            ${{ needs.sdk-tck-regression-panel.outputs.failure-mode || 'none' }}
            ${{ needs.mirror-node-regression-panel.outputs.failure-mode || 'none' }}
            ${{ needs.json-rpc-relay-regression-panel.outputs.failure-mode || 'none' }}
            ${{ needs.block-node-regression-panel.outputs.failure-mode || 'none' }}
            ${{ needs.migration-testing-panel.outputs.failure-mode || 'none' }}
```

**Review by andrewb1269 — CHANGES_REQUESTED** [review]

File: .github/workflows/zxc-block-node-regression.yaml [hunk 12]
```diff
@@ -429,6 +436,20 @@ jobs:
         run: |
           kind delete cluster -n ${{ env.SOLO_CLUSTER_NAME }}
 
+      - name: Initialize GitHub Job
+        uses: pandaswhocode/initialize-github-job@c392c3e699995ab6030915302d2fc9bffb1e861f # v1.1.0
+        with:
+          checkout: "true"
+          checkout-token: "${{ secrets.GITHUB_TOKEN }}"
+
```
Comment (andrewb1269) [inline .github/workflows/zxc-block-node-regression.yaml:428]: I believe the repo is checked out in step `Prepare Runner` at line 66.
```suggestion
```

File: .github/workflows/zxc-jrs-regression.yaml [hunk 13]
```diff
@@ -546,3 +553,17 @@ jobs:
           cat "${HOME}/release-resources.log"
           echo
           echo "<<<<<<<<<<<<<<<<<< End Diagnostic Log >>>>>>>>>>>>>>>>>>"
+
+      - name: Initialize GitHub Job
+        uses: pandaswhocode/initialize-github-job@c392c3e699995ab6030915302d2fc9bffb1e861f # v1.1.0
+        with:
+          checkout: "true"
+          checkout-token: "${{ secrets.GITHUB_TOKEN }}"
```
Comment (andrewb1269) [inline .github/workflows/zxc-jrs-regression.yaml]: Is the repo checked out in a previous step? If so, this checkout can be removed.

File: .github/workflows/zxc-json-rpc-relay-regression.yaml [hunk 14]
```diff
@@ -193,6 +200,20 @@ jobs:
         run: |
           kind delete cluster -n ${{ env.SOLO_CLUSTER_NAME }}
 
+      - name: Initialize GitHub Job
+        uses: pandaswhocode/initialize-github-job@c392c3e699995ab6030915302d2fc9bffb1e861f # v1.1.0
+        with:
+          checkout: "true"
+          checkout-token: "${{ secrets.GITHUB_TOKEN }}"
+
```
Comment (andrewb1269) [inline .github/workflows/zxc-json-rpc-relay-regression.yaml:192]: Repo is already checked out in the `Prepare Runner` step, so this Init GH Job step can be removed.
```suggestion
```

File: .github/workflows/zxc-mirror-node-regression.yaml [hunk 15]
```diff
@@ -162,6 +169,20 @@ jobs:
         run: |
           kind delete cluster -n ${{ env.SOLO_CLUSTER_NAME }}
 
+      - name: Initialize GitHub Job
+        uses: pandaswhocode/initialize-github-job@c392c3e699995ab6030915302d2fc9bffb1e861f # v1.1.0
+        with:
+          checkout: "true"
+          checkout-token: "${{ secrets.GITHUB_TOKEN }}"
+
```
Comment (andrewb1269) [inline .github/workflows/zxc-mirror-node-regression.yaml:161]: Repo is already checked out in `Prepare Runner` step at line 69, so this checkout can be removed.
```suggestion
```

**Review by andrewb1269 — CHANGES_REQUESTED** [review]

File: .github/workflows/zxc-jrs-regression.yaml [hunk 16]
```diff
@@ -546,3 +553,23 @@ jobs:
...
+        if: ${{ always() }}
+        working-directory: ${{ github.workspace }}
+        run: |
+          echo "checking for the script"
+          echo "Current working directory: $(pwd)"                                  # Print current directory (should be ${GITHUB_WORKSPACE})
+          if [ -e ".github/workflows/support/scripts/set-failure-mode.sh" ]; then   # Might need to set the working dirctory here and in the next step `${{ github.workspace }}/<path>`
+            echo "File exists"
+          else
+            echo "File does not exist"
+          fi
```
Comment (andrewb1269) [inline .github/workflows/zxc-jrs-regression.yaml]:
```suggestion
```

**Review by andrewb1269 — CHANGES_REQUESTED** [review]
Suggest removing a debugging step in `.github/workflows/zxc-jrs-regression.yaml`.

**Review by andrewb1269 — COMMENTED** [review]
LGTM, would like the XTS run [here](https://github.com/hiero-ledger/hiero-consensus-node/actions/runs/27358774360) to pass before I give final approval.

**Review by andrewb1269 — APPROVED** [review]
